x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/ChunkedTrainedModelPersister.java

```diff
@@ -0,0 +1,208 @@
+ */
+
+package org.elasticsearch.xpack.ml.dataframe.process;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.Version;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.LatchedActionListener;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+import org.elasticsearch.license.License;
+import org.elasticsearch.xpack.core.ml.dataframe.DataFrameAnalyticsConfig;
+import org.elasticsearch.xpack.core.ml.dataframe.analyses.Classification;
+import org.elasticsearch.xpack.core.ml.dataframe.analyses.Regression;
+import org.elasticsearch.xpack.core.ml.inference.TrainedModelConfig;
+import org.elasticsearch.xpack.core.ml.inference.TrainedModelInput;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+import org.elasticsearch.xpack.core.security.user.XPackUser;
+import org.elasticsearch.xpack.ml.dataframe.process.results.TrainedModelDefinitionChunk;
+import org.elasticsearch.xpack.ml.extractor.ExtractedField;
+import org.elasticsearch.xpack.ml.extractor.ExtractedFields;
+import org.elasticsearch.xpack.ml.extractor.MultiField;
+import org.elasticsearch.xpack.ml.inference.modelsize.ModelSizeInfo;
+import org.elasticsearch.xpack.ml.inference.persistence.TrainedModelDefinitionDoc;
+import org.elasticsearch.xpack.ml.inference.persistence.TrainedModelProvider;
+import org.elasticsearch.xpack.ml.notifications.DataFrameAnalyticsAuditor;
+
+import java.time.Instant;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toList;
+
+public class ChunkedTrainedModelPersister {
+
+    private static final Logger LOGGER = LogManager.getLogger(ChunkedTrainedModelPersister.class);
+    private final TrainedModelProvider provider;
+    private final AtomicReference<String> currentModelId;
+    private final DataFrameAnalyticsConfig analytics;
+    private final DataFrameAnalyticsAuditor auditor;
+    private final Consumer<Exception> failureHandler;
+    private final ExtractedFields extractedFields;
+    private volatile boolean readyToStoreNewModel = true;
+
+    public ChunkedTrainedModelPersister(TrainedModelProvider provider,
+                                        DataFrameAnalyticsConfig analytics,
+                                        DataFrameAnalyticsAuditor auditor,
+                                        Consumer<Exception> failureHandler,
+                                        ExtractedFields extractedFields) {
+        this.provider = provider;
+        this.currentModelId = new AtomicReference<>("");
```

Comment:
why not initialize it empty?

Reply:
it is empty?

Reply:
well, ... I mean `new AtomicReference<>()`